File: 04-distributed/01-partial-failure-and-the-ambiguous-result/nodejs/ambiguous_result.js
// Layer 4 · Topic 1 — the third outcome, in Node.js.
//
// WHAT THIS DEMONSTRATES
//   Node is the sharpest version of this topic for one reason: `fetch` has no
//   default timeout. Not a long one -- none. A dependency that accepts your
//   connection and then says nothing holds your handler until the OS gives up on
//   the socket, which on macOS and Linux is minutes. Phase 0 measures that
//   directly against a server that is deliberately silent.
//
//   Phases 1 and 2 then run the same five faults through the same client with
//   the same timeout, differing only in retry policy: phase 1 retries on any
//   error, phase 2 retries only errors that prove the request never landed.
//
// WHAT TO LOOK FOR
//   1. Phase 0: the request is still outstanding when the watchdog fires. Nothing
//      in the fetch call said "wait forever" -- that is just the default.
//   2. The `AbortSignal` note under phase 0. AbortSignal.timeout is the right fix
//      for the hang, but it is a deadline on the whole operation, so it reports
//      one TimeoutError whether it fired during the TCP handshake or halfway
//      through reading a response body. Hard errors like ECONNREFUSED still come
//      through on `cause`; timeouts -- the common case -- do not say which phase
//      they hit. That is why phases 1 and 2 use node:http, which does.
//   3. Phase 1's duplicate charges vs phase 2's, and the ambiguity that survives
//      both.
//
// Zero dependencies: node:net for the fault server, node:http for the client.

'use strict';

const net = require('node:net');
const http = require('node:http');

const CLIENT_TIMEOUT_MS = 300;   // the deadline the caller is willing to wait
const SLOW_RESPONSE_MS = 1000;   // deliberately longer than CLIENT_TIMEOUT_MS
const NO_TIMEOUT_WATCHDOG_MS = 3000;
const REQUESTS_PER_MODE = 4;
const MAX_ATTEMPTS = 3;

const MODES = [
  ['ok', 'no fault'],
  ['slow', 'commits, then replies after the client has given up'],
  ['hang', 'commits, then never replies at all (blackhole)'],
  ['reset', 'commits, then RSTs the connection'],
  ['crash_after_commit', 'commits, then dies before writing a byte of response'],
  ['refused', 'nothing is listening; the request provably never landed'],
];

// --- server-side truth ------------------------------------------------------

const ledger = [];              // every accepted charge, duplicates included

function startLedgerServer() {
  const held = [];              // `hang` sockets, destroyed at shutdown
  const server = net.createServer((socket) => {
    socket.once('data', (buf) => {
      const line = buf.toString('latin1').split('\r\n')[0];
      const path = line.split(' ')[1] || '';
      const [, , mode, chargeId] = path.split('/');

      const reply = (status = '200 OK') => {
        const body = JSON.stringify({ charge_id: chargeId });
        socket.end(
          `HTTP/1.1 ${status}\r\nContent-Type: application/json\r\n` +
          `Content-Length: ${Buffer.byteLength(body)}\r\nConnection: close\r\n\r\n${body}`
        );
      };

      switch (mode) {
        case 'ok':
          ledger.push(chargeId);
          reply();
          break;
        case 'slow':
          ledger.push(chargeId);
          setTimeout(reply, SLOW_RESPONSE_MS);
          break;
        case 'hang':
          ledger.push(chargeId);
          held.push(socket);            // accepted, committed, never answered
          break;
        case 'reset':
          ledger.push(chargeId);
          // destroy() on a socket with unread data sends RST rather than FIN.
          socket.resetAndDestroy();
          break;
        case 'crash_after_commit':
          // The case no timeout tuning can fix: durable work, dead reporter.
          ledger.push(chargeId);
          socket.destroy();
          break;
        default:
          reply('400 Bad Request');
      }
    });
    socket.on('error', () => { /* RSTs we caused ourselves */ });
  });
  server.held = held;
  return server;
}

// --- classification ---------------------------------------------------------
//
// The line is not "connection errors are safe, timeouts are not". It is "did we
// get far enough to have put the request bytes on the wire".
function classify(err) {
  const code = err && err.code;
  if (code === 'ECONNREFUSED' || code === 'EHOSTUNREACH' || code === 'ENOTFOUND') {
    return ['SAFE', code];            // never left this machine
  }
  if (code === 'ETIMEDOUT' && err.phase === 'connect') {
    return ['SAFE', 'connect ETIMEDOUT'];
  }
  if (err && err.phase === 'response') {
    return ['AMBIGUOUS', 'response timeout'];
  }
  if (code === 'ECONNRESET') {
    return ['AMBIGUOUS', 'ECONNRESET'];
  }
  if (err && err.message === 'socket hang up') {
    return ['AMBIGUOUS', 'socket hang up'];   // server closed before responding
  }
  // Anything unrecognised is ambiguous. Defaulting the other way is how
  // duplicate charges get shipped.
  return ['AMBIGUOUS', code || (err && err.message) || 'unknown'];
}

// One request with an explicit response deadline. node:http rather than fetch
// because it separates the connect phase from the response phase, and that
// separation *is* the safe/unsafe decision.
function request(port, path) {
  return new Promise((resolve) => {
    const req = http.request(
      { host: '127.0.0.1', port, path, agent: false },
      (res) => {
        res.resume();
        res.on('end', () => resolve(['SUCCESS', String(res.statusCode)]));
      }
    );
    let connected = false;
    req.on('socket', (socket) => {
      socket.on('connect', () => { connected = true; });
    });
    req.setTimeout(CLIENT_TIMEOUT_MS, () => {
      const err = new Error('timeout');
      err.code = 'ETIMEDOUT';
      err.phase = connected ? 'response' : 'connect';
      req.destroy(err);
    });
    req.on('error', (err) => {
      if (err.code === 'ETIMEDOUT' && err.phase === undefined) {
        err.phase = connected ? 'response' : 'connect';
      }
      resolve(classify(err));
    });
    req.end();
  });
}

// --- phase 0: the Node-specific headline ------------------------------------

async function phaseZero(port) {
  console.log('');
  console.log('  phase 0 — fetch() with no timeout, against a server that never replies');
  const started = Date.now();
  const url = `http://127.0.0.1:${port}/charge/hang/phase0-0`;

  const controller = new AbortController();
  const inFlight = fetch(url, { signal: controller.signal })
    .then(() => 'resolved')
    .catch((e) => `rejected: ${e.name}`);

  const watchdog = new Promise((r) =>
    setTimeout(() => r('STILL PENDING'), NO_TIMEOUT_WATCHDOG_MS));

  const outcome = await Promise.race([inFlight, watchdog]);
  const waited = Date.now() - started;
  console.log(`  after ${waited}ms the fetch is: ${outcome}`);
  console.log('  Nothing in that call asked to wait forever. That is the default,');
  console.log('  and every one of these pins a request slot and a socket.');
  controller.abort();
  await inFlight;

  // Why the obvious fix is not enough on its own.
  const connErr = await fetch(`http://127.0.0.1:${CLOSED_PORT}/x`,
    { signal: AbortSignal.timeout(CLIENT_TIMEOUT_MS) }).catch((e) => e);
  const readErr = await fetch(`http://127.0.0.1:${port}/charge/hang/phase0-1`,
    { signal: AbortSignal.timeout(CLIENT_TIMEOUT_MS) }).catch((e) => e);
  console.log('');
  console.log('  AbortSignal.timeout fixes the hang. What it does not give you:');
  console.log(`    connect to a closed port : ${connErr.name}` +
    `${connErr.cause ? ' (cause ' + connErr.cause.code + ')' : ''}`);
  console.log(`    silent server, mid-read  : ${readErr.name}`);
  console.log('  A hard refusal still identifies itself through `cause`. A timeout');
  console.log('  does not say which phase it fired in -- the same TimeoutError means');
  console.log('  "the handshake never finished" (safe) and "I sent the request and');
  console.log('  gave up waiting" (not safe). Phases 1 and 2 use node:http, which');
  console.log('  keeps the connect phase and the response phase apart.');
}

// --- phases 1 and 2 ---------------------------------------------------------

async function runPhase(tag, name, note, port, retryAmbiguous) {
  const before = ledger.length;
  const perMode = [];

  console.log('');
  console.log(`  ${name}`);
  console.log(`  ${note}`);
  console.log(`  ${'fault'.padEnd(20)}${'client verdict'.padEnd(34)}` +
    `${'attempts'.padStart(9)}${'ledger rows'.padStart(12)}`);

  for (const [mode] of MODES) {
    const modeBefore = ledger.length;
    const verdicts = new Map();
    let attempts = 0;
    const targetPort = mode === 'refused' ? CLOSED_PORT : port;

    for (let i = 0; i < REQUESTS_PER_MODE; i++) {
      const chargeId = `${tag}-${mode}-${i}`;
      let verdict, label;
      for (let a = 0; a < MAX_ATTEMPTS; a++) {
        attempts++;
        [verdict, label] = await request(targetPort, `/charge/${mode}/${chargeId}`);
        if (verdict === 'SUCCESS') break;
        if (verdict === 'SAFE') continue;          // provably safe: try again
        if (retryAmbiguous) continue;              // the bug, made explicit
        break;                                     // correct: stop, escalate
      }
      const key = `${verdict}(${label})`;
      verdicts.set(key, (verdicts.get(key) || 0) + 1);
    }

    const rows = ledger.length - modeBefore;
    const summary = [...verdicts].map(([v, n]) => `${n}x ${v}`).join(', ');
    console.log(`  ${mode.padEnd(20)}${summary.padEnd(34)}` +
      `${String(attempts).padStart(9)}${String(rows).padStart(12)}`);
    perMode.push(verdicts);
  }

  const written = ledger.slice(before);
  const counts = new Map();
  for (const id of written) counts.set(id, (counts.get(id) || 0) + 1);
  const duplicates = [...counts.values()].reduce((s, n) => s + (n > 1 ? n - 1 : 0), 0);
  const unresolved = perMode.reduce((s, v) => s +
    [...v].filter(([k]) => k.startsWith('AMBIGUOUS')).reduce((a, [, n]) => a + n, 0), 0);

  console.log(`  ledger rows written this phase : ${written.length}`);
  console.log(`  DUPLICATE CHARGES              : ${duplicates}` +
    '   <- created by this client\'s retries');
  console.log(`  unresolved ambiguous outcomes  : ${unresolved}` +
    '   <- caller cannot tell whether these happened');
  return { duplicates, unresolved };
}

// --- main -------------------------------------------------------------------

let CLOSED_PORT = 0;

function findClosedPort() {
  return new Promise((resolve) => {
    const probe = net.createServer();
    probe.listen(0, '127.0.0.1', () => {
      const { port } = probe.address();
      probe.close(() => resolve(port));
    });
  });
}

async function main() {
  const server = startLedgerServer();
  await new Promise((r) => server.listen(0, '127.0.0.1', r));
  const port = server.address().port;
  CLOSED_PORT = await findClosedPort();

  console.log('='.repeat(78));
  console.log('Layer 4 · Topic 1 — partial failure and the ambiguous result (Node.js)');
  console.log('='.repeat(78));
  console.log(`  ledger        : 127.0.0.1:${port}  (in-process, holds server-side truth)`);
  console.log(`  closed port   : 127.0.0.1:${CLOSED_PORT}  (for the connect-refused case)`);
  console.log(`  client timeout: ${CLIENT_TIMEOUT_MS}ms   slow response: ${SLOW_RESPONSE_MS}ms   ` +
    `max attempts: ${MAX_ATTEMPTS}`);

  await phaseZero(port);
  const naive = await runPhase(
    'p1',
    'phase 1 — retry on any error',
    'the `catch (e) { retry() }` most Node codebases have',
    port, true);
  const fixed = await runPhase(
    'p2',
    'phase 2 — retry only provably-safe errors',
    'ECONNREFUSED and connect-phase timeouts are retried; everything else escalates',
    port, false);

  console.log('');
  console.log('-'.repeat(78));
  console.log(`  duplicate charges    phase 1: ${String(naive.duplicates).padEnd(6)}` +
    `phase 2: ${fixed.duplicates}`);
  console.log(`  unresolved ambiguity phase 1: ${String(naive.unresolved).padEnd(6)}` +
    `phase 2: ${fixed.unresolved}`);
  console.log('');
  console.log('  The fix removes the duplicates the client was causing. It does not');
  console.log('  remove the ambiguity -- nothing at this layer can. Topic 2 is what');
  console.log('  makes retrying an ambiguous outcome safe.');

  for (const s of server.held) s.destroy();
  server.close();
}

main();

File: 04-distributed/01-partial-failure-and-the-ambiguous-result/python/ambiguous_result.py
"""
Layer 4 · Topic 1 — the third outcome, in Python (httpx).

WHAT THIS DEMONSTRATES
  A local call has two outcomes. A remote call has three, and the third --
  "I do not know whether it happened" -- is not an error you can engineer away.
  This program runs a real ledger server in-process so there is a *server-side
  truth* to compare the client's belief against, then injects five faults and
  shows the diff.

  Phase 1 retries on any exception, which is what `except httpx.HTTPError:`
  produces in almost every codebase. Phase 2 retries only the errors that prove
  the request never landed. Same faults, same load, both phases measured.

WHAT TO LOOK FOR
  1. Phase 1's "duplicate charges" count. Every one of those was created by the
     client's own retry, not by the fault.
  2. Phase 2's duplicate count, and the "unresolved ambiguous" count next to it.
     The duplicates go to zero; the ambiguity does not. That residue is
     irreducible at this layer and is the entire reason Topic 2 exists.
  3. `connect refused` is the only fault where a retry is provably safe -- the
     bytes never left the machine. Everything else is a guess.

WHY httpx: the exception taxonomy is the lesson. ConnectError/ConnectTimeout
mean the request provably never landed; ReadTimeout/RemoteProtocolError/ReadError
mean it may well have. Most code catches httpx.HTTPError, which is the parent of
all of them, and that single line is the bug.
"""
from __future__ import annotations

import socket
import struct
import sys
import threading
import time
from collections import Counter

try:
    import httpx
except ImportError:
    sys.exit("needs httpx:  python3 -m pip install httpx")

CLIENT_TIMEOUT = 0.3      # seconds; the deadline the caller is willing to wait
SLOW_RESPONSE = 1.0       # seconds; deliberately longer than CLIENT_TIMEOUT
REQUESTS_PER_MODE = 4
MAX_ATTEMPTS = 3

# The five faults, plus a baseline. `commits` records whether the server writes
# the charge to its ledger *before* the fault fires -- which is the whole point:
# four of these five commit and then fail to tell the caller.
MODES = [
    ("ok", "no fault"),
    ("slow", "commits, then replies after the client has given up"),
    ("hang", "commits, then never replies at all (blackhole)"),
    ("reset", "commits, then RSTs the connection"),
    ("crash_after_commit", "commits, then dies before writing a byte of response"),
    ("refused", "nothing is listening; the request provably never landed"),
]


class Ledger:
    """Server-side truth. Every accepted charge is appended, duplicates and all."""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self.rows: list[str] = []

    def commit(self, charge_id: str) -> None:
        with self._lock:
            self.rows.append(charge_id)

    def snapshot(self) -> list[str]:
        with self._lock:
            return list(self.rows)


class LedgerServer:
    """Minimal HTTP/1.1 over raw sockets.

    Raw sockets rather than http.server because two of the faults (an RST, and a
    close with no response at all) are things a WSGI-shaped handler cannot
    express. Being able to inject them precisely is worth the forty lines.

    Composition rather than subclassing threading.Thread on purpose: Thread
    already owns attribute names like `_handle` on 3.13, and silently shadowing
    one of them fails at runtime inside the thread where you will not see it.
    """

    def __init__(self, ledger: Ledger) -> None:
        self.ledger = ledger
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self.sock.bind(("127.0.0.1", 0))
        self.sock.listen(128)
        self.port = self.sock.getsockname()[1]
        self.running = True
        self._held: list[socket.socket] = []   # `hang` connections, closed at shutdown
        self._thread = threading.Thread(target=self._accept_loop, daemon=True)

    def start(self) -> None:
        self._thread.start()

    def _accept_loop(self) -> None:
        while self.running:
            try:
                conn, _ = self.sock.accept()
            except OSError:
                return
            threading.Thread(target=self._serve, args=(conn,), daemon=True).start()

    def _serve(self, conn: socket.socket) -> None:
        try:
            conn.settimeout(5.0)
            request = conn.recv(4096).decode("latin-1", "replace")
            if not request:
                conn.close()
                return
            path = request.split(" ", 2)[1]           # /charge/<mode>/<charge_id>
            _, _, mode, charge_id = path.split("/", 3)

            if mode == "ok":
                self.ledger.commit(charge_id)
                self._reply(conn, charge_id)
            elif mode == "slow":
                self.ledger.commit(charge_id)
                time.sleep(SLOW_RESPONSE)
                self._reply(conn, charge_id)
            elif mode == "hang":
                self.ledger.commit(charge_id)
                self._held.append(conn)             # never replied to, never closed
                return
            elif mode == "reset":
                self.ledger.commit(charge_id)
                # SO_LINGER with a zero timeout makes close() send RST, not FIN.
                # This is what a peer that panics or a middlebox that gives up
                # looks like from the client side.
                conn.setsockopt(socket.SOL_SOCKET, socket.SO_LINGER,
                                struct.pack("ii", 1, 0))
                conn.close()
                return
            elif mode == "crash_after_commit":
                # The case no timeout tuning can ever fix: the work is durable
                # and the only process that knew about it is gone.
                self.ledger.commit(charge_id)
                conn.close()
                return
            else:
                self._reply(conn, charge_id, status="400 Bad Request")
        except Exception:
            pass
        finally:
            if conn not in self._held:
                try:
                    conn.close()
                except OSError:
                    pass

    @staticmethod
    def _reply(conn: socket.socket, charge_id: str, status: str = "200 OK") -> None:
        body = f'{{"charge_id":"{charge_id}"}}'.encode()
        head = (
            f"HTTP/1.1 {status}\r\n"
            f"Content-Type: application/json\r\n"
            f"Content-Length: {len(body)}\r\n"
            f"Connection: close\r\n\r\n"
        ).encode()
        conn.sendall(head + body)

    def shutdown(self) -> None:
        self.running = False
        for held in self._held:
            try:
                held.close()
            except OSError:
                pass
        try:
            self.sock.close()
        except OSError:
            pass


# --- the classification that is the whole point -----------------------------
#
# SAFE means the request provably never reached the server, so a retry cannot
# duplicate work. AMBIGUOUS means it may have: the connection was established
# and the request bytes were written, and after that the client learns nothing.
#
# Note where the line falls. It is NOT "connection errors are safe, timeouts are
# not". It is "did we get far enough to have sent the request".
SAFE_TO_RETRY = (
    httpx.ConnectError,      # ECONNREFUSED / EHOSTUNREACH: never left the host
    httpx.ConnectTimeout,    # TCP handshake never completed
)
AMBIGUOUS = (
    httpx.ReadTimeout,       # request sent; we gave up waiting for the answer
    httpx.WriteTimeout,      # partially sent; the server may have a full copy
    httpx.RemoteProtocolError,   # server disconnected without a response
    httpx.ReadError,         # ECONNRESET mid-read
    httpx.PoolTimeout,       # never acquired a connection -- also safe, see note
)


def classify(exc: Exception) -> tuple[str, str]:
    """Return (verdict, short label). Verdict is SAFE, AMBIGUOUS or SUCCESS."""
    if isinstance(exc, httpx.PoolTimeout):
        # Pedantically safe: we never got a connection, so nothing was sent.
        # Kept separate because it is a *client* saturation signal, not a
        # dependency failure, and conflating the two hides an outage.
        return "SAFE", "PoolTimeout"
    if isinstance(exc, SAFE_TO_RETRY):
        return "SAFE", type(exc).__name__
    if isinstance(exc, AMBIGUOUS):
        return "AMBIGUOUS", type(exc).__name__
    # Anything unrecognised must be treated as ambiguous. Defaulting the other
    # way is how duplicate charges get shipped.
    return "AMBIGUOUS", type(exc).__name__


def attempt(client: httpx.Client, url: str) -> tuple[str, str]:
    try:
        response = client.get(url)
        return ("SUCCESS", str(response.status_code))
    except Exception as exc:  # noqa: BLE001 - classification is the subject
        return classify(exc)


def run_phase(name: str, port: int, retry_ambiguous: bool, note: str) -> dict:
    """One full pass over every fault mode.

    retry_ambiguous=True is the naive client: any exception, try again.
    retry_ambiguous=False only retries errors that prove nothing was sent.
    """
    ledger = SERVER.ledger
    before = len(ledger.snapshot())
    timeout = httpx.Timeout(
        connect=CLIENT_TIMEOUT, read=CLIENT_TIMEOUT,
        write=CLIENT_TIMEOUT, pool=CLIENT_TIMEOUT,
    )
    per_mode = {}

    print()
    print(f"  {name}")
    print(f"  {note}")
    print(f"  {'fault':<20} {'client verdict':<34} {'attempts':>9} {'ledger rows':>12}")

    with httpx.Client(timeout=timeout, limits=httpx.Limits(max_connections=200)) as client:
        for mode, _description in MODES:
            mode_before = len(ledger.snapshot())
            attempts = 0
            verdicts: Counter[str] = Counter()

            for i in range(REQUESTS_PER_MODE):
                charge_id = f"{name[:2]}-{mode}-{i}"
                target_port = port if mode != "refused" else CLOSED_PORT
                url = f"http://127.0.0.1:{target_port}/charge/{mode}/{charge_id}"

                for _ in range(MAX_ATTEMPTS):
                    attempts += 1
                    verdict, label = attempt(client, url)
                    if verdict == "SUCCESS":
                        break
                    if verdict == "SAFE":
                        continue                    # provably safe: try again
                    if retry_ambiguous:
                        continue                    # the bug, made explicit
                    break                           # correct: stop, escalate
                verdicts[f"{verdict}({label})"] += 1

            mode_rows = len(ledger.snapshot()) - mode_before
            summary = ", ".join(f"{n}x {v}" for v, n in verdicts.most_common())
            print(f"  {mode:<20} {summary:<34} {attempts:>9} {mode_rows:>12}")
            per_mode[mode] = verdicts

    rows = ledger.snapshot()[before:]
    counts = Counter(rows)
    duplicates = sum(n - 1 for n in counts.values() if n > 1)
    unresolved = sum(
        n for verdicts in per_mode.values()
        for v, n in verdicts.items() if v.startswith("AMBIGUOUS")
    )
    print(f"  {'':<20} {'':<34} {'':>9} {'':>12}")
    print(f"  ledger rows written this phase : {len(rows)}")
    print(f"  DUPLICATE CHARGES              : {duplicates}"
          f"   <- created by this client's retries")
    print(f"  unresolved ambiguous outcomes  : {unresolved}"
          f"   <- caller cannot tell whether these happened")
    return {"duplicates": duplicates, "unresolved": unresolved, "rows": len(rows)}


def find_closed_port() -> int:
    """A port nothing is listening on, so connect() gets ECONNREFUSED."""
    s = socket.socket()
    s.bind(("127.0.0.1", 0))
    port = s.getsockname()[1]
    s.close()
    return port


SERVER: LedgerServer
CLOSED_PORT: int


def main() -> None:
    global SERVER, CLOSED_PORT
    ledger = Ledger()
    SERVER = LedgerServer(ledger)
    SERVER.start()
    CLOSED_PORT = find_closed_port()

    print("=" * 78)
    print("Layer 4 · Topic 1 — partial failure and the ambiguous result (Python/httpx)")
    print("=" * 78)
    print(f"  ledger        : 127.0.0.1:{SERVER.port}  (in-process, holds server-side truth)")
    print(f"  closed port   : 127.0.0.1:{CLOSED_PORT}  (for the connect-refused case)")
    print(f"  client timeout: {CLIENT_TIMEOUT}s   slow response: {SLOW_RESPONSE}s   "
          f"max attempts: {MAX_ATTEMPTS}")

    naive = run_phase(
        "phase 1 — retry on any exception",
        SERVER.port,
        retry_ambiguous=True,
        note="what `except httpx.HTTPError: retry` actually does",
    )
    fixed = run_phase(
        "phase 2 — retry only provably-safe errors",
        SERVER.port,
        retry_ambiguous=False,
        note="ConnectError/ConnectTimeout are retried; everything else escalates",
    )

    print()
    print("-" * 78)
    print(f"  duplicate charges   phase 1: {naive['duplicates']:<6} "
          f"phase 2: {fixed['duplicates']}")
    print(f"  unresolved ambiguity phase 1: {naive['unresolved']:<6} "
          f"phase 2: {fixed['unresolved']}")
    print()
    print("  The fix removes the duplicates the client was causing. It does not")
    print("  remove the ambiguity -- nothing at this layer can, because the")
    print("  information needed is on the far side of the thing that broke.")
    print("  Making a retry of an ambiguous outcome *safe* is Topic 2.")
    SERVER.shutdown()


if __name__ == "__main__":
    main()
